packages/kbn-generate-csv/src/generate_csv_esql.ts

```diff
@@ -0,0 +1,221 @@
+          ? this.job.columns.filter((column) => responseColumns.includes(column))
+          : responseColumns;
+
+      const rows = rawResponse.values.map((row) => zipObject(responseColumns, row));
+
+      const header =
+        Array.from(visibleColumns).map(this.escapeValues(settings)).join(settings.separator) + '\n';
+      builder.tryAppend(header);
+
+      await this.generateRows(visibleColumns, rows, builder, settings);
+    } catch (err) {
+      this.logger.error(err);
+      if (err instanceof esErrors.ResponseError) {
+        if ([401, 403].includes(err.statusCode ?? 0)) {
+          reportingError = new AuthenticationExpiredError();
+          warnings.push(i18nTexts.authenticationError.partialResultsMessage);
+        } else {
+          warnings.push(i18nTexts.esErrorMessage(err.statusCode ?? 0, String(err.body)));
+        }
+      } else {
+        warnings.push(i18nTexts.unknownError(err?.message ?? err));
+      }
+    }
+
+    return {
+      content_type: CONTENT_TYPE_CSV,
+      csv_contains_formulas: this.csvContainsFormulas && !escapeFormulaValues,
+      max_size_reached: this.maxSizeReached,
+      metrics: {
+        csv: { rows: this.csvRowCount },
+      },
+      warnings,
+      error_code: reportingError?.code,
+    };
+  }
+
+  /*
+   * Format a Datatable into rows of CSV content
+   */
+  private async generateRows(
+    columns: string[],
+    rows: Array<Record<string, unknown>>,
+    builder: MaxSizeStringBuilder,
+    settings: CsvExportSettings
+  ) {
+    this.logger.debug(`Building ${rows.length} CSV data rows`);
+    for (const dataTableRow of rows) {
+      if (this.cancellationToken.isCancelled()) {
+        break;
+      }
+
+      /*
+       * Intrinsically, generating the rows is a synchronous process. Awaiting
+       * on a setImmediate call here partititions what could be a very long and
+       * CPU-intenstive synchronous process into an asychronous process. This
+       * give NodeJS to process other asychronous events that wait on the Event
+       * Loop.
+       *
+       * See: https://nodejs.org/en/docs/guides/dont-block-the-event-loop/
+       *
+       * It's likely this creates a lot of context switching, and adds to the
+       * time it would take to generate the CSV. There are alternatives to the
+       * chosen performance solution:
+       *
+       * 1. Partition the synchronous process with fewer partitions, by using
+       * the loop counter to call setImmediate only every N amount of rows.
+       * Testing is required to see what the best N value for most data will
+       * be.
+       *
+       * 2. Use a C++ add-on to generate the CSV using the Node Worker Pool
+       * instead of using the Event Loop
+       */
+      await new Promise(setImmediate);
+
+      const rowDefinition: string[] = [];
+      const escape = this.escapeValues(settings);
+
+      for (const column of columns) {
+        rowDefinition.push(escape(`${dataTableRow[column]}`));
```

Comment:
Maybe as you had it was intentional, but this looks the same as:
```suggestion
        rowDefinition.push(escape(dataTableRow[column]));
```

Reply:
safe typecasting from unknown to string + if there is anything something then string, would be safely converted to string